Allow opening relative paths returned by the assistant in the user's IDE

fluiddot opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to 8984-gh-Automattic/dotcom-forge.

## Proposed Changes

- Expand `CodeBlock` component by adding logic to infer the content type to render. There are two options:
  1. `language`: When the content contains code of a specific programming language. This will render the component `LanguageBlock` which might display actions like running the code.
  2. `file`: When the content represents a potential file path. This will render the component `File` that might allow opening the file in an IDE if the file exists.
  3. Other content: This will be rendered without any extra UI.
- Add unit tests to cover the new functionality.

![image](https://github.com/user-attachments/assets/2b409251-7e97-458c-8eff-505c7ebecfb1)

![image](https://github.com/user-attachments/assets/9120eca5-2789-4f54-a48e-8b8efff46015)


> [!IMPORTANT]
> I'm using the color `a8c-blueberry` when the file can be opened. However, there might be contrast issues depending on the code format that comes in the response. @matt-west I'd appreciate if you could take a look and share thoughts/feedback on this. Thanks 🙇 ! **UPDATE:** This has been solved in https://github.com/Automattic/studio/pull/549/commits/764e2865c2e45037130aa9f4603736d54292824f.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

> [!NOTE]
> I experienced performance issues when having several Assistant messages and typing in the input field. This will be addressed in a separate PR.

- Run the app with the command `npm start`.
- Log in to WPCOM with a A8C account.
- Navigate to the Assistant tab.
- Submit a prompt that will produce a response with references to files (e.g. `How can I enable debug logs in the site?`).
- Observe that the Assistant's response contains file path references.
- Observe that some of the file path references can be clicked (only the ones associated with existing files in the site).
- Click on one that references a file.
- Observe that the file is opened in an installed IDE (VSCode or PHPStorm).
- Click on one that references a directory.
- Observe that the directory is opened in the Finder/File Explorer.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

fluiddot committed:
Add IPC handler to obtain the absolute path of site files

fluiddot committed:
Allow opening file paths referenced in Assistant responses in an IDE

fluiddot committed:
Add unit tests to cover file path content Assistant responses

matt-west commented:
Thanks for the ping @fluiddot!

I think we need an affordance to communicate links that can be opened in the user’s IDE. I’ll work on a design.

matt-west commented:
@fluiddot Is it necessary to link folders here? (i.e. `wp-content`)

Linking individual files (`wp-config.php`) has a lot of value, but will users what to open a folder in their IDE?

fluiddot commented:
> @fluiddot Is it necessary to link folders here? (i.e. `wp-content`)
> 
> Linking individual files (`wp-config.php`) has a lot of value, but will users what to open a folder in their IDE?

Good point, probably it's not necessary to include folders in this functionality. I presume users would like to edit code files but not open entire folders in the IDE. I'll remove this from the PR.

matt-west commented:
@fluiddot Adding an edit icon helps to make it clearer that these files can be opened in an IDE. [Figma](https://www.figma.com/design/RToz6tIuQ7nlZrikBte4GU/Studio?node-id=4330-337406&t=7W9AIAd4Dk2GBDQr-4)

![image](https://github.com/user-attachments/assets/87dce91d-a2cf-4996-9d75-7e3224dae8a0)

I don’t think we should highlight files referenced inside code blocks. (For example, the `/wp-content/debug.log` path shown in the code comment here.) User's should be able to easily copy the raw code and if we start manipulating it by adding icons and links, it could have unintended consequences.

![image](https://github.com/user-attachments/assets/ff6f4416-c298-41da-a229-619f59497606)

fluiddot committed:
Add WordPress Blue 70 color to Tailwind config

fluiddot committed:
Add icon to `FileBlock` component

fluiddot committed:
Update styles of `FileBlock` component

fluiddot committed:
Only allow opening files in IDE from Assistant response

We no longer allow folders as it doesn't provide much value when opening them in an IDE.

fluiddot committed:
Fix unit tests

fluiddot commented:
@matt-west I've updated the styles with your input, thanks 🙇 ! Here's a video capture:

https://github.com/user-attachments/assets/9becf431-ff5e-40d4-b92a-d8c2a81d4b20

I noticed that the color used for the hover state is quite similar to the regular one, but I wouldn't consider it a blocker for the PR. Let me know if you rather update it or keep it as-is.

matt-west commented:
Thanks @fluiddot. This looks great!

> I noticed that the color used for the hover state is quite similar to the regular one, but I wouldn't consider it a blocker for the PR.

I agree that it’s quite close. This needs to be addressed in the design system then we can circle back and fix it in Studio. Links elsewhere in the app have the same issue.

wojtekn commented:
Nice improvement @fluiddot ! I tested it, and the basic flow worked for me. I've received an answer on how to add constants and the link allowed me to easily open the exact file I needed.

![Screenshot 2024-09-20 at 12 29 14](https://github.com/user-attachments/assets/5f2d1573-4dd7-4c14-897d-1ef87c0d2c09)

The only thing I noticed is that it doesn't always open the file in the context of the project:
- If I already added a site directory as a project in PhpStorm, and I had it open before I closed the IDE, it opens the file in the correct project (expected)
- If I already added a site directory as a project in PhpStorm, and I had another project opened before I closed the IDE, it opens the file in the correct project (expected)
- If I haven't added a site directory as a project in PhpStorm yet, and I closed the IDE, it opens the file in the light editor and asks me if I want to open it in the project. (expected)
- If I opened another project in PhpStorm, it opens the file in this project (not expected)

Do we have any control over that to improve the last case? Or is it how the IDE integration works?

> Good point, probably it's not necessary to include folders in this functionality. I presume users would like to edit code files but not open entire folders in the IDE. I'll remove this from the PR.

It could be still helpful to allow opening directories in IDE, for example, when users ask Assistant to write plugin or theme, it would provide path where they should place the file.

matt-west commented:
> It could be still helpful to allow opening directories in IDE, for example, when users ask Assistant to write plugin or theme, it would provide path where they should place the file.

Thanks for highlighting this. It feels more of a niche use-case, but we can support it. I'd still avoid linking paths within code blocks though.

fluiddot commented:
> [...] The only thing I noticed is that it doesn't always open the file in the context of the project:
> 
> * If I already added a site directory as a project in PhpStorm, and I had it open before I closed the IDE, it opens the file in the correct project (expected)
> * If I already added a site directory as a project in PhpStorm, and I had another project opened before I closed the IDE, it opens the file in the correct project (expected)
> * If I haven't added a site directory as a project in PhpStorm yet, and I closed the IDE, it opens the file in the light editor and asks me if I want to open it in the project. (expected)
> * If I opened another project in PhpStorm, it opens the file in this project (not expected)
> 
> Do we have any control over that to improve the last case? Or is it how the IDE integration works?

Good catch @wojtekn! As far as I know, this is handled by the IDE but I think we could make a project (pointing to the site folder) opens first. This way we ensure that the file will be opened in the context of the site folder. I'll take a look and update the PR.

> > Good point, probably it's not necessary to include folders in this functionality. I presume users would like to edit code files but not open entire folders in the IDE. I'll remove this from the PR.
> 
> It could be still helpful to allow opening directories in IDE, for example, when users ask Assistant to write plugin or theme, it would provide path where they should place the file.

Opening a folder could be interesting but not sure if it would be helpful in the IDE. However, we could let the user open it in the Finder, WDYT?

matt-west commented:
Finder makes more sense to me, but I happy to follow your judgement on this one.

If we go with Finder, we should use the folder icon next to the link rather than an edit icon.

fluiddot commented:
> Finder makes more sense to me, but I happy to follow your judgement on this one.
> 
> If we go with Finder, we should use the folder icon next to the link rather than an edit icon.

Good idea, I'll apply this update too. Thanks Matt 🙇 !

fluiddot committed:
Add `openFileInIDE` IPC handler

fluiddot committed:
Use `openFileInIDE` IPC handler in Assistant code block component

fluiddot committed:
Open WP directory references in Assistant messages with Finder

fluiddot committed:
Update unit tests of Assistant code block

fluiddot commented:
@wojtekn I've updated the logic to open the file in IDE in order to ensure it's opened within the site project context. I tested in VSCode and works as expected.

- Site project already created
    - VSCode opened ✅ 
    - VSCode closed ✅ 
- Site project not created
    - VSCode opened ✅ 
    - VSCode closed ✅ 
- Another project opened in VSCode ✅ 

However, I couldn't test PHPStorm IDE because I completed the 30-day trial, would you mind taking a look? Thanks 🙇 !

fluiddot commented:
> > Finder makes more sense to me, but I happy to follow your judgement on this one.
> > If we go with Finder, we should use the folder icon next to the link rather than an edit icon.
> 
> Good idea, I'll apply this update too. Thanks Matt 🙇 !

I also updated how directories are handled. Now they are opened in the Finder and render a folder icon next to the path.

![image](https://github.com/user-attachments/assets/e6649ad0-5053-44bb-a2b1-07b67b50e373)

wojtekn commented:
> Thanks for highlighting this. It feels more of a niche use-case, but we can support it.

It doesn't seem like a niche to me, as writing custom code like plugins is one of the tasks we would like Assistant to be good with.

wojtekn commented:
> Good catch @wojtekn! As far as I know, this is handled by the IDE but I think we could make a project (pointing to the site folder) opens first. This way we ensure that the file will be opened in the context of the site folder. I'll take a look and update the PR.

Thanks for the change. I tested it, and it opens the site folder as a project in one PhpStorm window and the `wp-config.php` in another project I had already opened, so it's not ideal yet.

> Opening a folder could be interesting but not sure if it would be helpful in the IDE. However, we could let the user open it in the Finder, WDYT?

It makes sense to open Finder, thanks. The instructions are much more convenient now:

![Screenshot 2024-09-23 at 12 58 41](https://github.com/user-attachments/assets/7e4ffa11-b3c3-4577-ae12-ab87c78915f4)

In the future, besides the "Copy" action, we could have another one that lets user to download or create the file.

fluiddot commented:
> Thanks for the change. I tested it, and it opens the site folder as a project in one PhpStorm window and the `wp-config.php` in another project I had already opened, so it's not ideal yet.

Oh 😞... Ok, I'll keep exploring how to solve it. Thanks for testing it 🙇 !

wojtekn commented:
> Oh 😞... Ok, I'll keep exploring how to solve it. Thanks for testing it 🙇 !

As far as I see, `phpstorm` deep links are not really documented. I played with the `project` and `file` params but with no luck. If it doesn't work, we can revert to the previous state, so it won't ideally work in this particular case.

fluiddot committed:
Revert try to open site project in PHPStorm

fluiddot commented:
> As far as I see, `phpstorm` deep links are not really documented. I played with the `project` and `file` params but with no luck. If it doesn't work, we can revert to the previous state, so it won't ideally work in this particular case.

Yep, I encountered the same limitation. I've reverted the change in https://github.com/Automattic/studio/pull/549/commits/2db235de9d4b1e20f022d4e4b97cee64cedbf27b.
